Added Markdown Editor

EOEboh opened the pull request:
> First thing, PLEASE READ THIS: [ReactPlay Code Review Checklist](https://github.com/reactplay/react-play/wiki/ReactPlay-Code-Review-Checklist)

# Description
I added a new play called Markdown Editor #430 .

It's a beginner-level play which displays input texts as markdown output. It utilizes a dependency called 'remarkable'.

My main motivation is to teach beginners the React library by making them take baby steps before tackling advanced concepts. 


- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

It works perfectly on localhost, netlify, vercel and mobile platforms.

# Checklist:

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

Sachin-chaurasiya commented:
Hey @EOEboh , Thanks for the PR.

In the code, I can see you are using `dangerouslySetInnerHTML` within the div tag.
```
<div className='output-div' dangerouslySetInnerHTML={{__html: md.render(text)}} >
```
Can you avoid this?
Or can we use [tui editor](https://ui.toast.com/tui-editor/) , which will provide both markdown and WYSIWYG mode?

Tui Editor is highly configurable and customisable, with plugin architecture it allows you to add your own plugin and also you can use an existing plugin to add an extra feature to the markdown editor.

<img width="1113" alt="image" src="https://user-images.githubusercontent.com/59080942/181879638-1ba19ad6-c4fa-4018-9e3d-ae190b8465f4.png">

cc : @atapas

EOEboh commented:
> Hey @EOEboh , Thanks for the PR.
> 
> In the code, I can see you are using `dangerouslySetInnerHTML` within the div tag.
> 
> ```
> <div className='output-div' dangerouslySetInnerHTML={{__html: md.render(text)}} >
> ```
> 
> Can you avoid this? Or can we use [tui editor](https://ui.toast.com/tui-editor/) , which will provide both markdown and WYSIWYG mode?
> 
> Tui Editor is highly configurable and customisable, with plugin architecture it allows you to add your own plugin and also you can use an existing plugin to add an extra feature to the markdown editor.
> 
> <img alt="image" width="1113" src="https://user-images.githubusercontent.com/59080942/181879638-1ba19ad6-c4fa-4018-9e3d-ae190b8465f4.png">
> 
> cc : @atapas

Thanks @Sachin-chaurasiya 
I would check it out.

Quick one: I did a `git reset` yesterday and I couldn't find my play on localhost after pushing changes. Do you have any suggestions how to fix it

EOEboh commented:
@atapas can you review this please?

atapas commented:
> > Hey @EOEboh , Thanks for the PR.
> > In the code, I can see you are using `dangerouslySetInnerHTML` within the div tag.
> > ```
> > <div className='output-div' dangerouslySetInnerHTML={{__html: md.render(text)}} >
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Can you avoid this? Or can we use [tui editor](https://ui.toast.com/tui-editor/) , which will provide both markdown and WYSIWYG mode?
> > Tui Editor is highly configurable and customisable, with plugin architecture it allows you to add your own plugin and also you can use an existing plugin to add an extra feature to the markdown editor.
> > <img alt="image" width="1113" src="https://user-images.githubusercontent.com/59080942/181879638-1ba19ad6-c4fa-4018-9e3d-ae190b8465f4.png">
> > cc : @atapas
> 
> Thanks @Sachin-chaurasiya I would check it out.
> 
> Quick one: I did a `git reset` yesterday and I couldn't find my play on localhost after pushing changes. Do you have any suggestions how to fix it

@EOEboh Do you still have the issue?

atapas commented:
@EOEboh Could you please resolve merge conflicts?

atapas reviewed (changes requested):
Please do not commit this file

src/plays/index.js

atapas commented on src/plays/markdown-editor/Readme.md:
```diff
@@ -0,0 +1,15 @@
+# Markdown Editor
+
```
Please add more details like,

- What someone will learn from this project?
- You have included an NPM, please mention that in the readme.

atapas commented on src/plays/markdown-editor/styles.css:
```diff
@@ -0,0 +1,26 @@
+/* enter stlyes here */
+*{
```
Do not override with *

atapas commented on src/plays/markdown-editor/styles.css:
```diff
@@ -0,0 +1,26 @@
+    
+  }
+  
+  h1, h3{
```
Do not override selectors.

atapas commented on src/plays/markdown-editor/styles.css:
```diff
@@ -0,0 +1,26 @@
+    box-sizing: border-box; 
+  }
+  
+  .play-details{
```
This is a common style across plays.. Please add a unique style and use that.

atapas commented on src/plays/markdown-editor/styles.css:
```diff
@@ -0,0 +1,26 @@
+  .play-details{
+    background-color: #724040;
+  }
+  .main{
```
This is has got higher chances of getting overridden. Please use a unique class name.

atapas reviewed (changes requested):
@EOEboh 

Provided my review comments

atapas commented on src/plays/index.js:
```diff
@@ -1,5 +1,4 @@
 // Do not modify the content of this file
```
@EOEboh as mentioned this file shouldn't be committed.

atapas commented on src/plays/markdown-editor/MarkdownEditor.js:
```diff
@@ -0,0 +1,47 @@
+
+  return (
+    <>
+      <div className="play-details">
```
Add a unique class name here along with `play-details`. Like 

```html
<div className="eboh-md-editor play-details">
```

Now use this one as a prefix to the styles in your CSS file. Like
```css
.eboh-md-editor.play-details {


} 

.eboh-md-editor .wrapper-div {

}
// Like the above.
```

atapas commented on src/plays/markdown-editor/MarkdownEditor.js:
```diff
@@ -0,0 +1,47 @@
+          
+        <h3 className='heading-3'> Preview</h3>
+
+          <div className='output-div' dangerouslySetInnerHTML={{__html: md.render(text)}} >
```
Do you need the `Remarkable`? As I understand, we are not highlighting any syntax here, so why do we need to treat the output as HTML, why not the plain text?

atapas commented on src/plays/markdown-editor/styles.css:
```diff
@@ -0,0 +1,36 @@
+    box-sizing: var(--box);
+  }
+  
+  .wrapper-div{
```
Please follow the structure I mentioned about the unique style name.

atapas commented on src/plays/markdown-editor/styles.css:
```diff
@@ -0,0 +1,36 @@
+/* enter stlyes here */
+:root {
```
Please do not override. These selectors are already overridden at the framework root level. So remove.

atapas reviewed (changes requested):
@EOEboh A few more change requests.

atapas commented:
@EOEboh 

Also please check this error. The build fails due to this



static/css/main.617be5a3.css from Css Minimizer plugin
--
08:04:26.826 | /vercel/path0/static/css/main.617be5a3.css:7170:27: Missed semicolon [webpack://./src/plays/markdown-editor/styles.css:11,4][static/css/main.617be5a3.css:7170,27]

EOEboh commented:
@atapas I have made corrections.
However, the conflicting `src/plays/index.js` keeps being pushed despite being in the `.gitignore` file.

Please can you resolve the conflict? Or tell me what to do.

EOEboh commented:
@koustov could you have a look, if it's good to go

atapas commented:
> @koustov could you have a look, if it's good to go

To me it's almost there. But you need to resolve the conflicts. Just revert the file content back to whatever is in the main.

atapas commented on src/plays/markdown-editor/styles.css:
```diff
@@ -0,0 +1,33 @@
+    font-family: system-ui, -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, Oxygen, Ubuntu, Cantarell, 'Open Sans', 'Helvetica Neue', sans-serif;
+    box-sizing: border-box;
+  }
+  
```
use `.md-editor` for rest of the styles too.

atapas commented:
@EOEboh Please take a look at my comments.

EOEboh commented:
> > @koustov could you have a look, if it's good to go
> 
> To me it's almost there. But you need to resolve the conflicts. Just revert the file content back to whatever is in the main.

What do you really mean here? Can you outline the steps?

koustov commented:
I have reviewed and it looks good.
@EOEboh a play got merged, so you need to resolve the conflicts in order to merge

koustov reviewed (approved):
lgtm

EOEboh commented:
> I have reviewed and it looks good. @EOEboh a play got merged, so you need to resolve the conflicts in order to merge

I have resolved it now. Please confirm

koustov commented:
> > I have reviewed and it looks good. @EOEboh a play got merged, so you need to resolve the conflicts in order to merge
> 
> I have resolved it now. Please confirm

Approved.

EOEboh commented:
> > > I have reviewed and it looks good. @EOEboh a play got merged, so you need to resolve the conflicts in order to merge
> > 
> > 
> > I have resolved it now. Please confirm
> 
> Approved.

Feels great to see this lol

koustov commented:
Hey @EOEboh , this looks good however do you feel like updating the UI a bit to make it more contemporary? 

You can refer 
- https://dillinger.io/
- https://markdownlivepreview.com/
- https://codebeautify.org/markdown-editor
- https://www.tutorialspoint.com/online_markdown_editor.php

EOEboh commented:
> Hey @EOEboh , this looks good however do you feel like updating the UI a bit to make it more contemporary?
> 
> You can refer
> 
> * https://dillinger.io/
> * https://markdownlivepreview.com/
> * https://codebeautify.org/markdown-editor
> * https://www.tutorialspoint.com/online_markdown_editor.php

Sure. I'd work on that

EOEboh commented:
@koustov and @atapas  I have modified the UI
I also added a download feature, to download the HTML text output as a pdf file

atapas commented:
Thanks for improving it @EOEboh 

Can you please change the text colour of the type here and preview the texts? They are barely readable with black font and brown background. This is a mandatory fix.

Other optional improvement suggestions:

- The boxes can be side by side, then up and down. They should be responsive so that on a small screen, they align column-wise.
- Can provide border-radius to the boxes.
- The Download button can be disabled if there are no texts.
- The cover image design can be better.

Please note that the above comments are optional for you to take care of, but if you do, it will help the play to perceive well. Also, if you think of getting it done and need any help or peer-programming, just let us know in the discord. One of us will work with you.

Thanks.

EOEboh commented:
> 

Okay, I'd work on it

EOEboh commented:
@atapas I have made corrections and implemented some of your suggestions

koustov commented:
> > 
> 
> @atapas Okay, I'd work on it. I just wrote an article about the play, is there a way I can edit the description, and link it?

You can create a link on your play

koustov commented:
> @atapas I have made corrections and implemented your suggestions

Hey @EOEboh , me and @atapas  were suggesting on similar lines. Few things are still pending
- A better cover page
- `Type Here` and `Preview` are in black on top of dark back ground which is not readable
- Disable download button when not text in `preview` box
- Make both horizontally aligned and make the most of real estate. something like 
![image](https://user-images.githubusercontent.com/7145967/183305053-a0e46b4f-d20c-4f57-9bf6-7db22a3140ff.png)


Could you please make the changes

EOEboh commented:
> > @atapas I have made corrections and implemented your suggestions
> 
> Hey @EOEboh , me and @atapas were suggesting on similar lines. Few things are still pending
> 
> * A better cover page
> * `Type Here` and `Preview` are in black on top of dark back ground which is not readable
> * Disable download button when not text in `preview` box
> * Make both horizontally aligned and make the most of real estate. something like
>   ![image](https://user-images.githubusercontent.com/7145967/183305053-a0e46b4f-d20c-4f57-9bf6-7db22a3140ff.png)
> 
> Could you please make the changes

I already implemented them 🙂

koustov commented:
> > > @atapas I have made corrections and implemented your suggestions
> > 
> > 
> > Hey @EOEboh , me and @atapas were suggesting on similar lines. Few things are still pending
> > 
> > * A better cover page
> > * `Type Here` and `Preview` are in black on top of dark back ground which is not readable
> > * Disable download button when not text in `preview` box
> > * Make both horizontally aligned and make the most of real estate. something like
> >   ![image](https://user-images.githubusercontent.com/7145967/183305053-a0e46b4f-d20c-4f57-9bf6-7db22a3140ff.png)
> > 
> > Could you please make the changes
> 
> I already implemented them 🙂

😮 then I am surprised because I didn't see it. Will check again

koustov commented:
@EOEboh , looks like latest build didn't get deployed to vercel, it still shows the thumbnail as 
![image](https://user-images.githubusercontent.com/7145967/183307954-a04b158f-5879-4c48-b41b-125a7037502d.png)

and the play as
![image](https://user-images.githubusercontent.com/7145967/183307963-8fad4b54-0942-4c11-a703-27d8bd2f00d1.png)

Will wait for @atapas to fix the vercel issue first

atapas commented:
@EOEboh much better looking, great work

@koustov, please conclude.

koustov commented:
UI is impressive @EOEboh. I will finish my review by EOD

EOEboh commented:
> UI is impressive @EOEboh. I will finish my review by EOD

Thank you @koustov and @atapas .. this means a lot to me.

koustov commented on package.json:
```diff
@@ -19,7 +19,9 @@
     "html-to-image": "^1.9.0",
-    "json-graphql-parser": "^0.1.7",
+    "html2canvas": "^1.4.1",
+    "json-graphql-parser": "^0.0.20",
```
Let's not downgrade the library.

koustov commented on src/plays/markdown-editor/Downloader.jsx:
```diff
@@ -0,0 +1,32 @@
+import React from 'react';
+
+// html2canvas for converting the md to png image
+import html2canvas from 'html2canvas';
```
Lets not use `html2canvas` as it brings the style as well

koustov replied:
![image](https://user-images.githubusercontent.com/7145967/183733853-b1497e21-c2bb-4e22-ae5a-7d612ca1fc35.png)

koustov commented on src/plays/markdown-editor/Downloader.jsx:
```diff
@@ -0,0 +1,32 @@
+import React from 'react';
+
+// html2canvas for converting the md to png image
+import html2canvas from 'html2canvas';
```
also, as its a png/jpg, one cannot copy the text from pdf

koustov commented on src/plays/markdown-editor/MarkdownEditor.js:
```diff
@@ -0,0 +1,68 @@
+          elementId='elementId'
+          />
+        </div>
+          <Downloader
```
Need to disable it when no text is rendered. I could download an empty document.

koustov commented on src/plays/markdown-editor/MarkdownEditor.js:
```diff
@@ -0,0 +1,68 @@
+            
+            </textarea>
+          
+            {/* <h3 className='md-editor heading-3'>
+        Preview
+      </h3> */}
```
Remove commented snippets

koustov commented on src/plays/markdown-editor/MarkdownEditor.js:
```diff
@@ -0,0 +1,68 @@
+          <Output 
+          md={md}
+          text={text}
+          elementId='elementId'
```
it's very generic id, I would suggest using more specific one like `mde-preview-box`

koustov commented on src/plays/markdown-editor/Readme.md:
```diff
@@ -0,0 +1,30 @@
+
+- User: EOEboh
+- Gihub Link: https://github.com/EOEboh
+- Blog: 
+- Video: 
```
Lets remove unused properties

koustov commented on src/plays/markdown-editor/styles.css:
```diff
@@ -0,0 +1,68 @@
+/* enter stlyes here */
```
Can you run `prettier` on this file? Looks like the alignments are not proper

koustov reviewed (changes requested):
Few observations

koustov commented:
@EOEboh is it ready for final review?

EOEboh commented:
> @EOEboh is it ready for final review?

@koustov I'm sorry I'm delaying, I was down these few days. I would work on the changes today. I apologize

EOEboh commented:
@koustov I fixed the issues.

koustov reviewed (approved):
lgtm

EOEboh commented:
> lgtm

@koustov I made a small change.
I think everything is set.

atapas commented:
@all-contributors please add @EOEboh for Code
